private/bufpkg/bufplugin/bufplugindocker/docker_test.go

```diff
@@ -119,3 +186,142 @@ func buildDockerPlugin(t testing.TB, dockerClient bufplugindocker.Client, docker
 	}
 	return response, err
 }
+
+// dockerServer allows testing some failure flows by simulating the responses to Docker CLI commands.
+type dockerServer struct {
+	httpServer    *httptest.Server
+	t             testing.TB
+	versionPrefix string
+	buildErr      error
+	pushErr       error
+	// protects builtImages
+	mu          sync.RWMutex
+	builtImages map[string]*builtImage
+}
+
+type builtImage struct {
+	tags []string
+}
+
+func newDockerServer(t testing.TB, version string) *dockerServer {
+	versionPrefix := "/v" + version
+	dockerServer := &dockerServer{
+		t:             t,
+		builtImages:   make(map[string]*builtImage),
+		versionPrefix: versionPrefix,
+	}
+	mux := http.NewServeMux()
+	mux.HandleFunc("/session", dockerServer.sessionHandler)
+	mux.HandleFunc(versionPrefix+"/build", dockerServer.buildHandler)
+	mux.HandleFunc(versionPrefix+"/images/", dockerServer.imagesHandler)
+	dockerServer.httpServer = httptest.NewServer(h2c.NewHandler(mux, &http2.Server{}))
+	t.Cleanup(dockerServer.httpServer.Close)
+	return dockerServer
+}
+
+func (d *dockerServer) sessionHandler(w http.ResponseWriter, r *http.Request) {
+	if _, err := io.Copy(io.Discard, r.Body); err != nil {
+		d.t.Error("failed to discard body:", err)
+	}
+	if r.Method != http.MethodPost {
+		http.NotFound(w, r)
+		return
+	}
+	if strings.EqualFold(r.Header.Get("Connection"), "upgrade") && r.ProtoMajor < 2 {
+		// Needed to allow h2c upgrade to proceed on the /session endpoint.
+		w.WriteHeader(http.StatusSwitchingProtocols)
+		return
+	}
+	w.WriteHeader(http.StatusNotFound)
+}
+
+func (d *dockerServer) buildHandler(w http.ResponseWriter, r *http.Request) {
+	if _, err := io.Copy(io.Discard, r.Body); err != nil {
+		d.t.Error("failed to discard body:", err)
+	}
+	w.WriteHeader(http.StatusOK)
+	if d.buildErr != nil {
+		d.writeError(w, d.buildErr)
+		return
+	}
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	imageID := stringid.GenerateRandomID()
+	d.builtImages[imageID] = &builtImage{tags: r.URL.Query()["t"]}
+}
+
+var imagePattern = regexp.MustCompile("^(?P<image>[^/]+/[^/]+/[^/:]+)(?::(?P<tag>[^/]+))?(?:/(?P<op>[^/]+))?$")
```

Comment:
Let's declare this at the top of the file - all global, un-exported `const` and `var` declarations should be at the top.